### ISSUE TITLE
Wrong dataLabels position when setting fixed x or y values.

### ISSUE THREAD

**Issue by Denyllon** [highcharts/highcharts#10137]
#### Expected behaviour
Data labels should always be aligned to the edge of which they would potentially get out.

#### Actual behaviour
When the 'dataLabels.x' or 'dataLabels.y' fixed values are set, and the data label is getting out of the plotArea, then its offset is set relative to point plot position instead of achieving the maximum possible value (closest to the edge which is going out of plot area boundaries).

Data label which has `y` position set to maximum possible (in this case `40px`) looks like below:
![screenshot 2019-02-14 at 12 33 29](https://user-images.githubusercontent.com/15111955/52784267-d0cd1c80-3054-11e9-8f6a-ed174846c7d8.png)

After setting it equal to `41px`, a problem is noticeable like below:
![screenshot 2019-02-14 at 12 35 32](https://user-images.githubusercontent.com/15111955/52784329-0c67e680-3055-11e9-869b-e0e72e467c5d.png)

If we additionally increase the `y` value to `50px` then we're able to see that there is a difference between the maximum available space and `y` value added to the dataLabel position, which is actually relative to point plot position:
![screenshot 2019-02-14 at 12 42 43](https://user-images.githubusercontent.com/15111955/52784666-2950e980-3056-11e9-9252-ffbd9c1add99.png)

#### Use case ####
The timeline series uses the data label's `options.x` and `options.y` parameters, in order to set its initial offset. 

#### Live demo with steps to reproduce
Live example: https://jsfiddle.net/zfL54bot/
Change the `dataLabels.y` value in order to see different effects.

_Internal Note:_
_Probably it's happening because of fact that the `Series.justifyDataLabel()` function always overrides the `options.x` and `options.y` values, when the data label exceeds the plot boundaries:_

https://github.com/highcharts/highcharts/blob/261b5e3f40fdc396118b6506ec7547adb8eba04d/js/parts/DataLabels.js#L1131

_Possible fix here could be incrementing/decrementing existing `y` or `x` value, for example:_
```
options.x -= off;
```

**Comment by TorsteinHonsi** [comment]
@Denyllon I'm reverting this because it broke a lot of the visual tests. The reason for the current behaviour (that it doesn't simply align to the edge of the plot area), is that it is designed not to cover up the point that the data label is aligned to. 

So if the data label is vertically aligned to show above the point, and it detects that it overflows the plot area, it is moved down below the point. Likewise if it overflows to the right, it is not only moved just in side the plot area, but rather realigned to the left of its point. You can see the difference here: http://jsfiddle.net/highcharts/f3nuq1g9/. With this fix, the data label would cover the point.

**Comment by Denyllon** [comment]
@TorsteinHonsi Right, I know that and that's why I fixed the way of aligning them, because that behaviour seems to me unnatural. If someone sets the data labels fixed position by defining `x` and `y` values, and they're moved to unexpected position by some reason (when they haven't got enough space on the chart), it can be misleading.

Actually, I've noticed that during implementing Timeline series. See what happening when resizing the window on this example: http://jsfiddle.net/BlackLabel/mznbg49a/

So, to conclude, we need to consider the other way of justifying data labels than existing one, but I think that should be changed in general.

**Comment by pawelfus** [comment]
I had the same issue with annotations + crop&overflow. That behaviour looks a bit strange, take a look: 

![ann-crop](https://user-images.githubusercontent.com/1453926/55069221-d1a99300-5083-11e9-839a-ae4e32828068.gif)


In stock tools, we decided to disable crop&overflow for basic-label annotation because of that. Can me make this optional somehow?

**Comment by Denyllon** [comment]
@pawelfus Furthermore, the same issue is noticeable when either data label's fixed `x` value and `draggable-points.js` are applied on Torstein's example.

http://jsfiddle.net/BlackLabel/q53darz6/1/ (drag any of points and move it to the left plot area boundary)

**Comment by TorsteinHonsi** [comment]
In summary, we need a solution that works for both cases.

**Comment by pawelfus** [comment]
Unstale
